### ISSUE TITLE
Coordinates retrieved from plugin are not always accurate.

### ISSUE THREAD

**Issue by Stepo45** [pmwisdom/cordova-background-geolocation-services#26]
Hello,

thx for your plugin. I have a problem with cofiguration, track isn't smooth, is there some recommended configuration for running or driving? The location updates data are very inaccurate.

My current config is:

desiredAccuracy: 0,
distanceFilter: 0,
interval: 9000,
fastestInterval: 5000,
useActivityDetection: true

Thx for your answer.

**Comment by pmwisdom** [comment]
I don't have any control over which GPS points get sent back via the Native API's. GPS inaccuracies can be caused by a bunch of different things, including your location. 

Can you elaborate on what you mean by "track isnt smooth".

What platform are you using (Android / iOS) ?
What are the accuracy readings that you get back in the location passed back through the location update callback?

**Comment by Stepo45** [comment]
Hi, below is a screenshot of my previous tracking. I think you see whats going on. (I was driving accross the "yellow" road (Hodoninska) towards the blue marker, and this is how my route got tracked.
Tested on Android (on more devices with same, or similar unaccurate results).

Accuracy readings coming in a while.

![screenshot_2016-06-14-19-32-54](https://cloud.githubusercontent.com/assets/19933535/16053334/ab660700-3268-11e6-9469-a52f8db48b88.png)

**Comment by pmwisdom** [comment]
If its device agnostic I'd like to say its probably that your hitting an area with bad GPS, but the accuracy readings will tell us if its the plugins problem or the GPS's problem.

**Comment by Stepo45** [comment]
Accuracy: unfortunately, I didn't store accuracy in this tracking. But currently, while moving around the building, I'm getting various numbers from 5 to 95. Maybe I should give it a try once again, and make a few kilometer driving to check it. I'm in a location with standard GPS signal, other applications are tracking my movement well.

**Comment by Stepo45** [comment]
OK, I made a test drive around a block, which gave me a bit better (but still not accurate) results, and the accuracy was like 6 (average). The highest number I received was 109.43, lowest was 5.

**Comment by pmwisdom** [comment]
The other applications your using might also be throwing away GPS positions with low accuracy (higher accuracy numbers mean less accurate). If you get coordinates back over 70 accuracy I would throw them away.

On android especially, the coordinates you get passed back are out of my hands, that is up to the device and android Location libraries, the configuration you have is the most accurate you can go on android. 

The most important being : 

```
desiredAccuracy: 0
```

Which translates to give me the most accurate results and put all the power you can into the GPS.

**Comment by Stepo45** [comment]
Yes, that's good idea to throw the inaccurate coordinates away. I'll try this right away and post you the results.

**Comment by Stepo45** [comment]
OK, I was able to get better results, by some tweaking of result coordinates data set. I'm not displaying inaccurate results, and moreover, I'm checking the distance between two adjacent coordinates. If they're too far away, I will not display the second one.
Haven't you thought about implementing "Location.distanceBetween()" inside the plugin?
(you'll not send coordinate, which is too far away from the previous one).

**Comment by pmwisdom** [comment]
I'm glad you were able to get better results!

There isn't really a reason to implement that in the native code since you can do so just as easily via JavaScript and everyone has different requirements from the plugin. I want to leave the plugin as open and broad as possible.

I would implement it if there was a good enough reason of why it couldn't / would be harder to do via JavaScript.

**Comment by sepideh68** [comment]
Hi pmwisdom and Stepo45 , I have the same problem when tracking, specially with android. Stepo45 can you show how you are checking the distance between two adjacent coordinates? 
The program that I am developing is Cordova base and it should support all the functionality same for both iOS and Android.
With iOS sometimes is it good and sometimes it is not working good. I have attached the same track with different result on iOS. here is my configure:
`desiredAccuracy: 0,
distanceFilter: 3,
debug: true,
interval: 1000,
fastestInterval: 1000,
useActivityDetection: true`

![sample2](https://cloud.githubusercontent.com/assets/14283034/16059083/70164a84-324f-11e6-901c-eef99e694668.jpg)

**Comment by pmwisdom** [comment]
@sepideh68 Its very simple. You can use the formula in this stack overflow.

http://stackoverflow.com/questions/27928/calculate-distance-between-two-latitude-longitude-points-haversine-formula

**Comment by sepideh68** [comment]
@pmwisdom  I am receiving almost accurate coordinates on iOS but it is not working on android. 
Is it because of config I am using?! If so why it is working good on iOS?! I am developing my app with ionic. here is the code I have put on ionic.Platform.ready function:

``` javascript
 ionic.Platform.ready(function(){
 cordova.plugins.backgroundMode.enable();
cordova.plugins.backgroundMode.onactivate = function () {
bgLocationServices.start();
$scope.isBackground = true;
           }
     cordova.plugins.backgroundMode.ondeactivate = function () {
    bgLocationServices.stop();  
    $scope.isBackground = false;
            }
    navigator.geolocation.getCurrentPosition(function(p){}); // it is required
    bgLocationServices =  window.plugins.backgroundLocationServices;

    //Congfigure Plugin
    bgLocationServices.configure({
               //Both
                 desiredAccuracy:      0, 
                 distanceFilter:       0, 
                 debug:                false, 
                 interval:             2000, 
                 fastestInterval:      1000, 
                 useActivityDetection: false,
                 notificationTitle: '', // customize the title of the notification
                 notificationText: '', //customize the text of the notification     
            });

    bgLocationServices.registerForLocationUpdates(function(location) {
        if($scope.started){
        var position = encapsulateLocation(location);
        showPosition2(position);
            }
      }
```

**Comment by pmwisdom** [comment]
@sepideh68 Please fix your styling and be more specific. Show me some coordinates on android vs iOS. 

Also why are you using the cordova backgroundMode plugin? It might be unwise to do so with this plugin since it starts another service in the background.

**Comment by sepideh68** [comment]
@pmwisdom I am using cordova backgroundMode plugin to enable tracking when background is activate and user started the geolocation. Does it have conflict with your plugin?! BTW I want to collect Gyro and Accelerometer data even if user is on backgroundMode. Here is the result of same route:

![randroid](https://cloud.githubusercontent.com/assets/14283034/16202806/4c7eddc6-36e5-11e6-877d-b519b0f471e2.png)
![riphone](https://cloud.githubusercontent.com/assets/14283034/16202807/4c8bfe5c-36e5-11e6-869a-5575f61a89eb.png)

**Comment by pmwisdom** [comment]
@sepideh68 My plugin has the effect of keeping your app alive for however long you are tracking with it, so you don't need the background mode plugin. I don't know if its causing conflict but I would remove it. 

Also, try starting the background plugin _before_ you transition to the background, and before the will resign callback.

**Comment by srsarcasmo** [comment]
This is good talk, I just released a compiled for android platform and I am using this plugin very well, I like the way it works; In ios i have an application that is not kept in the background and I will use this plugin for this, besides I will use the callback to execute a javascript function and not to use the data that the plugin returns, so it could be my function I will leave them down

**Comment by srsarcasmo** [comment]
`function timer_nativo(){
document.addEventListener("deviceready", onDeviceReady, false);
function onDeviceReady() {
navigator.geolocation.getCurrentPosition(function() {
// alert("el gps esta trabajando");
}, function(error) {
   alert(error);
});

var bgLocationServices =  window.plugins.backgroundLocationServices;
bgLocationServices.configure({
     //Both android and ios
     desiredAccuracy: 10, //Precisión deseada de las actualizaciones de ubicación (menor significa más exacta pero más consumo de batería)
     distanceFilter: 0, //(Medidores) Hasta qué punto debe moverse desde el último punto para activar una actualización de ubicación
     debug: true, // <--Habilitar para mostrar las indicaciones visuales cuando recibe una actualización de la ubicación de fondo
     interval: 120000, //(Milisegundos) Intervalo solicitado entre actualizaciones de ubicación.
     useActivityDetection: false, // Utiliza la detección de Actividad para apagar los gps cuando todavía está (mejora enormemente la duración de la batería)

     //Android Only
     notificationTitle: 'ABpro Systems', // title of the notification
     notificationText: 'Geolocalizacion', //text of the notification
     fastestInterval: 120000 // <--(Milisegundos) Intervalo más rápido que su aplicación o servidor puede manejar actualizaciones
});

//Registrar una devolución de llamada para las actualizaciones de ubicación, aquí es donde los objetos de ubicación se enviarán en segundo plano
bgLocationServices.registerForLocationUpdates(function(location) {
     autoUpload2();
  //   alert("We got an BG Update" + JSON.stringify(location));
  // se ejecuta cuando se detecta una nueva ubicacion
}, function(err) {
     alert("Error: Didnt get an update ForLocationUpdates", err);
});

bgLocationServices.registerForActivityUpdates(function(activities) {
   //  alert("We got an activity update" + activities);
}, function(err) {
   //  alert("Error: Something went wrong", err);
});

bgLocationServices.start();


///later, to stop
//bgLocationServices.stop();

}
}

`

**Comment by srsarcasmo** [comment]
In ios i can not compile, i do not understand the reason why the service has not been modified in native code; I'm using iphone 6 target;

**Comment by srsarcasmo** [comment]
![android debug](https://cloud.githubusercontent.com/assets/20671621/22993735/b8075e22-f389-11e6-8eb9-fde08e2f8158.png)

**Comment by srsarcasmo** [comment]
@pmwisdom 
Because ios does not compile the app when I add the plugin to work?

help me.. @sepideh68

**Comment by pmwisdom** [comment]
@srsarcasmo  This is a completely unrelated issue. Please do not spam the other issues.
